saw-core/src/SAWCore/Rewriter.hs

```diff
@@ -868,6 +871,131 @@ rewriteSharedTermTypeSafe sc ss t0 =
 
+data Convertibility = AllRules | ConvertibleRulesOnly
+
+rewriteSharedTermConvertibility :: forall a. Ord a => SharedContext -> Simpset a -> Term -> IO (Set a, Term)
+rewriteSharedTermConvertibility sc ss t0 =
+    do cache <- newCache
+       let ?cache = cache
+       setRef <- newIORef mempty
+       let ?annSet = setRef
+       t <- rewriteAll AllRules t0
+       anns <- readIORef setRef
+       pure (anns, t)
+
+  where
+
+    rewriteAll :: (?cache :: Cache IO TermIndex Term, ?annSet :: IORef (Set a)) => Convertibility -> Term  -> IO Term
+
+    rewriteAll convertibleFlag STApp{ stAppIndex = tidx, stAppTermF = tf } =
+        useCache ?cache tidx (rewriteTermF convertibleFlag tf >>= scTermF sc >>= rewriteTop convertibleFlag)
+
+    rewriteTermF :: (?cache :: Cache IO TermIndex Term, ?annSet :: IORef (Set a)) =>
+                    Convertibility -> TermF Term -> IO (TermF Term)
+    rewriteTermF convertibleFlag tf =
+        case tf of
+          FTermF ftf -> FTermF <$> rewriteFTermF convertibleFlag ftf
+          App e1 e2 ->
+              do t1 <- scTypeOf sc e1
+                 case unwrapTermF t1 of
+                   -- If type of e1 is not a dependent type, we can use any rule to rewrite e2
+                   -- otherwise, we only rewrite using convertible rules
+                   -- This prevents rewriting e2 from changing type of @App e1 e2@.
+                   Pi x _ t
+                     | IntSet.notMember (vnIndex x) (freeVars t) ->
+                         App <$> rewriteAll convertibleFlag e1 <*> rewriteAll convertibleFlag e2
+                   _ -> App <$> rewriteAll convertibleFlag e1 <*> rewriteAll ConvertibleRulesOnly e2
+                   -- could compute WHNF of t1 to see if it's a thing that doesn't match Pi but behaves like Pi
+          Lambda pat t e -> Lambda pat t <$> rewriteAll convertibleFlag e -- pat is x or varnames etc, t types (so don't rewrite), e is body
+          Constant{}     -> return tf
+          Variable{}     -> return tf
+          Pi x t1 t2 -> Pi x <$> rewriteAll convertibleFlag t1 <*> rewriteAll convertibleFlag t2
+
+    rewriteFTermF :: (?cache :: Cache IO TermIndex Term, ?annSet :: IORef (Set a)) =>
+                     Convertibility -> FlatTermF Term -> IO (FlatTermF Term)
+    rewriteFTermF convertibleFlag ftf =
+        case ftf of
+          UnitValue        -> return ftf
+          UnitType         -> return ftf
+          PairValue{}      -> traverse (rewriteAll convertibleFlag) ftf
+          PairType{}       -> traverse (rewriteAll convertibleFlag) ftf
+          PairLeft{}       -> traverse (rewriteAll convertibleFlag) ftf
+          PairRight{}      -> traverse (rewriteAll convertibleFlag) ftf
+
+          -- NOTE: we don't rewrite arguments of constructors, datatypes, or
+          -- recursors because of dependent types, as we could potentially cause
+          -- a term to become ill-typed
```

Comment:
This comment is from the original `rewriteSharedTermTypeSafe`, but it is outdated and should be removed. We don't have special term constructors for data constructor applications, datatype applications, or recursor applications anymore (#2358). These are all covered by the `App` case now, and don't require any special cases in the rewriter.